forge/test/mlir/llama/tests/test_llama_prefil.py

```diff
@@ -0,0 +1,78 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+import pytest
+from transformers import LlamaTokenizer
+
+import forge
+from test.mlir.llama.utils.utils import load_model
+from forge.op.eval.common import compare_with_golden
+
+def prefil_on_cpu(model, input_ids):
+    with torch.no_grad():
+        transformer_outputs = model.model(
+            input_ids=input_ids,  # Pass the entire updated sequence
+        )
+        hidden_states = transformer_outputs.last_hidden_state
+    return hidden_states
+
+def decode_on_cpu(model, tokenizer, input_ids, hidden_states, max_new_tokens):
+    for i in range(max_new_tokens):
+        # Use only the hidden state of the last token in the sequence
+        next_token_logits = model.lm_head(hidden_states[:, -1, :])
+
+        # Get the next token ID
+        next_token_id = torch.argmax(next_token_logits, dim=-1)
+
+        # Stop if the EOS token is generated
+        if next_token_id == tokenizer.eos_token_id:
+            break
+
+        # Update input_ids with the new token
+        input_ids = torch.cat([input_ids, next_token_id.unsqueeze(0)], dim=-1)
+
+        # Run the model again to get the new hidden state
+        hidden_states = prefil_on_cpu(model, input_ids)
+    
+    return input_ids, hidden_states
+
+@pytest.mark.xfail()
+def test_llama_prefil_on_device_decode_on_cpu():
+    """
+    This function tests the inference of the Llama 3B model split into two parts:
+    - The first part is the prefilling of the model on the device.
+    - The second part is the decoding of the model on the CPU.
+    """
+    # Load Llama 3B model and tokenizer
+    model_path = "openlm-research/open_llama_3b"
+    model = load_model(model_path, use_cache=False, return_dict=True)
+    tokenizer = LlamaTokenizer.from_pretrained(model_path)
+
+    # Prepare input sentence
+    prompt = "Q: What is the largest animal?\nA:"
+    input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+    
+    # This is the part of the model needed for prefill; model without the last Linear layer (lm_head)
```

Comment:
As we discussed offline, prefill stage is defined as the stage where we prepare hidden states and KV cache to be able to do efficient decode. In our case, putting lm head inside prefill stage would mean we have created one token output before we started the decode. Therefore, leaving it like this because is more inline with intention of prefill and decode...